File: boj/java/src/mst/Boj1976.java
package mst;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Boj1976 {
	static int find(int[] p, int a) {
		if (p[a] != a) p[a] = find(p, p[a]);
		return p[a];
	}

	static void union(int[] p, int a, int b) {
		int pa = find(p, a);
		int pb = find(p, b);

		if (pa < pb) p[pb] = pa;
		else p[pa] = pb;
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		int m = Integer.parseInt(br.readLine());

		int[] p = new int[n];

		for (int i = 0; i < n; i++) p[i] = i;

		for(int i = 0; i < n; i++) {
			int[] arr = Arrays.stream(br.readLine().split(" "))
					.mapToInt(Integer::parseInt)
					.toArray();

			for (int j = 0; j < n; j++) {
				if (arr[j] == 1) union(p, i, j);
			}
		}

		int[] path = Arrays.stream(br.readLine().split(" "))
				.mapToInt(Integer::parseInt)
				.toArray();
		int set = find(p, path[0] - 1);
		String answer = "YES";

		for (int i = 1; i < m; i++) {
			if (find(p, path[i] - 1) != set) {
				answer = "NO";
				break;
			}
		}
		System.out.println(answer);
	}
}

File: boj/python/mst/boj1976.py
def solution():
    return


def find(p, a):
    if p[a] != a:
        p[a] = find(p, p[a])
    return p[a]


def union(p, a, b):
    pa = find(p, a)
    pb = find(p, b)

    if pa < pb:
        p[pb] = pa
    else:
        p[pa] = pb


n = int(input())
m = int(input())
answer = "YES"

parent = [i for i in range(n)]
for i in range(n):
    adj = list(map(int, input().split()))
    for j in range(n):
        if adj[j]:
            union(parent, i, j)

path = list(map(int, input().split()))
_set = find(parent, path[0])

for i in range(m):
    if find(parent, path[i] - 1) != _set:
        answer = "NO"
        break
print(answer)
